cloud/google/gceproviderconfig/v1alpha1/register.go

```diff
@@ -44,14 +52,62 @@ func addKnownTypes(scheme *runtime.Scheme) error {
 	return nil
 }
 
-func NewSchemeAndCodecs() (*runtime.Scheme, *serializer.CodecFactory, error) {
+func NewScheme() (*runtime.Scheme, error) {
 	scheme := runtime.NewScheme()
 	if err := AddToScheme(scheme); err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 	if err := gceproviderconfig.AddToScheme(scheme); err != nil {
-		return nil, nil, err
+		return nil, err
+	}
+	return scheme, nil
+}
+
+func NewCodec() (*GCEProviderConfigCodec, error) {
+	scheme, err := NewScheme()
+	if err != nil {
+		return nil, err
+	}
+	codecFactory := serializer.NewCodecFactory(scheme)
+	encoder, err := newEncoder(&codecFactory)
+	if err != nil {
+		return nil, err
+	}
+	codec := GCEProviderConfigCodec{
+		encoder: encoder,
+		decoder: codecFactory.UniversalDecoder(SchemeGroupVersion),
+	}
+	return &codec, nil
+}
+
+func (codec *GCEProviderConfigCodec) DecodeFromProviderConfig(providerConfig clusterv1.ProviderConfig) (*GCEProviderConfig, error) {
```

Comment:
cc @mkjelland 
I think we're going to move to where we have separate config for clusters and machines, and also we're going to have provider status for both of them too. Would it be possible for this API to take a `runtime.Object` to decode into instead of returning a specific type?

Also, could `EncodeToProviderConfig` take a `runtime.Object` as well?

This probably doesn't need to be done in this PR.

Reply:
Happy to take that up in a follow up, more targetted PR.

Reply:
I can take care of this in my PR to create the separate config types. I was thinking it might be nicer to have 2 methods, that way they can return the correct type. Otherwise the caller has to cast the return value to the type they want, right? Or did you have something else in ming @krousey?

Reply:
I was thinking of just having an output parameter of type `runtime.Object`. That way you could do something like

```go
func (codec *GCEProviderConfigCodec) DecodeProviderConfigInto(providerConfig runtime.RawExtension, out runtime.Object) error {...}

var blah GCEProviderConfig
err := c.DecodeProviderConfigInto(m.Spec.ProviderConfig.Value, &blah)
```